pkg/aggregator/demultiplexer.go

```diff
@@ -0,0 +1,365 @@
+	"github.com/DataDog/datadog-agent/pkg/epforwarder"
+	"github.com/DataDog/datadog-agent/pkg/forwarder"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// DemultiplexerInstance is a shared global demultiplexer instance.
+// Initialized by InitAndStartAgentDemultiplexer or InitAndStartServerlessDemultiplexer,
+// could be nil otherwise.
+//
+// The plan is to deprecated this global instance at some point.
+var demultiplexerInstance Demultiplexer
+
+var demultiplexerInstanceMu sync.Mutex
+
+// Demultiplexer is composed of multiple samplers (check and time/dogstatsd)
+// a shared forwarder, the event platform forwarder, orchestrator data buffers
+// and other data that need to be sent to the forwarders.
+// DemultiplexerOptions let you configure which forwarders have to be started.
+type Demultiplexer interface {
+	// General
+
+	Run()
+	Stop(flush bool)
+
+	// Aggregation API
+
+	// AddTimeSamples adds time samples processed by the DogStatsD server into a time sampler pipeline.
+	// The MetricSamples should have their hash computed.
+	AddTimeSamples(sample []metrics.MetricSample)
+	// AddCheckSample adds check sample sent by a check from one of the collectors into a check sampler pipeline.
+	AddCheckSample(sample metrics.MetricSample)
+	// FlushAggregatedData flushes all the aggregated data from the samplers to
+	// the serialization/forwarding parts.
+	FlushAggregatedData(start time.Time, waitForSerializer bool)
+	// Aggregator returns an aggregator that anyone can use. This method exists
+	// to keep compatibility with existing code while introducing the Demultiplexer,
+	// however, the plan is to remove it anytime soon.
+	//
+	// Deprecated.
+	Aggregator() *BufferedAggregator
+	// Serializer returns a serializer that anyone can use. This method exists
+	// to keep compatibility with existing code while introducing the Demultiplexer,
+	// however, the plan is to remove it anytime soon.
+	//
+	// Deprecated.
+	Serializer() serializer.MetricSerializer
+
+	// Senders API, mainly used by collectors/checks
+
+	GetSender(id check.ID) (Sender, error)
+	SetSender(sender Sender, id check.ID) error
+	DestroySender(id check.ID)
+	GetDefaultSender() (Sender, error)
+	ChangeAllSendersDefaultHostname(hostname string)
+	cleanSenders()
+}
+
+// AgentDemultiplexer is the demultiplexer implementation for the main Agent
+type AgentDemultiplexer struct {
+	m sync.Mutex
+
+	// options are the options with which the demultiplexer has been created
+	options    DemultiplexerOptions
+	aggregator *BufferedAggregator
+	dataOutputs
+	*senders
+}
+
+// DemultiplexerOptions are the options used to initialize a Demultiplexer.
+type DemultiplexerOptions struct {
+	ForwarderOptions              *forwarder.Options
+	UseNoopEventPlatformForwarder bool
+	NoEventPlatformForwarder      bool
+	NoOrchestratorForwarder       bool
```

Comment:
Also for the PR #10011 on which we need to add another Forwarder, I was think on how modify the function `serializer.NewSerializer()` to have something more generic. see my comment: https://github.com/DataDog/datadog-agent/pull/10011#discussion_r765675196

it would be nice if we can have the same capability with the demultiplexer. 

if in `DemultiplexerOptions` we can expose `[]SerializerOptions` we can have something more generic

```golang 
type DemultiplexerOptions struct {
  //...
  serializerOptions []serializer.SerializerOptions
  //...
}

func ForwarderFactoryFunc func() *forwarder.DefaultForwarder
func (opts *DemultiplexerOptions) WithOrchestratorForwarderFunc(orchFunc ForwarderFactoryFunc) *DemultiplexerOptions {
  orchestratorOptionsFunc = func(s *serializer.Serializer) {
    s.SetOrchestratorForwarder(orchFunc())
  }
  opts.SerializerOptions = append(opts.serializerOptions, orchestratorOptionsFunc)
  return opts
}

func (opts *DemultiplexerOptions) WithEventPlatform(orchFunc ForwarderFactoryFunc) *DemultiplexerOptions {
  //...
  opts.serializerOptions = append(opts.serializerOptions, serializeEventPlatformOptions)
  return opts
}
```
then 

```golang
func InitAndStartAgentDemultiplexer(options DemultiplexerOptions, hostname string) *AgentDemultiplexer {
  // ...
  sharedSerializer := serializer.NewSerializer(sharedForwarder, options.serializerOptions...)
}

```

depending of where we use the `Demultiplexer` we can use or not the function `WithOrchestratorForwarderFunc()`

```golang
opts := aggregator.DefaultDemultiplexerOptions(nil)
opts.WithOrchestratorForwarderFunc(orch.NewOrchestratorForwarder)

```

Reply:
I addressed the first comment.

About the second one,  since this change is fairly transversal even if I agree that it would be a logic evolution to address #10011, I'm trying to add as less as possible, just adapting to the existing codebase for now.
I feel it could be addressed in either:
  * an additional PR to add this feature
  * the #10011 PR

WDYT?

Reply:
> I feel it could be addressed in either:
  - an additional PR to add this feature 
  - the feat: Collect container lifecycle events #10011 PR


2 cents: I think it's out of scope for #10011 , #10011 shouldn't change how the aggregator/serialiser/forwarder work.

Reply:
@ahmed-mez it is already introducing a new forwarder and adapting the NewSerializer signature, which is not much more than the helpers/options changes mentioned by Cedric (i.e. if the demux was already merged, you may have consider doing the Cedric's improvements just because it is better than just adding the change without refactoring); but don't get me wrong, I agree, this change would be better in its own PR.

I'll do it in the current one.

Reply:
thanks! the plan was to rebase and merge  #10011  after this PR gets merged anyway.

Reply:
Not strong opinion, 
Indeed it can be added in a future PR. 
My point is: when I see own many files, tests need to be modify to add a new forwarder in the PR #10011,  I really think updating the `NewSerializer` interface is worth doing.